Themis specification: keys and keypairs (#25)
cossacklabs/product-docs: themis/spec/symmetric-keys.md

Themis currently generates 256-bit keys by default (i.e., 32 bytes).

See also our [key management guide](/themis/crypto-theory/key-management/)
for advice on securing the key after you have generated it.

### Example

Here is a snippet in Go, illustrating how to generate a good symmetric key:

```go
package main

import (
	"crypto/rand" // Use a cryptograpic RNG.
	"fmt"
)

func main() {
	// Generate keys of sufficient length.
	const keyBytes = 32
	key := make([]byte, keyBytes)
	_, err := rand.Read(key)
	if err != nil {
		// Not handling RNG failure is a common critical error.
		// If errors are ignored, you might return a predictable key,
		// such as "all zeros" initial value of the byte array.
		panic("failed to generate a key")
	}
	fmt.Printf("✅ symmetric key: %x\n", key)
}
```

## Passphrases

Alternatively, Secure Cell can work with passphrases.
From the cryptographic point of view, a passphrase is essentially a symmetric key with lower entropy per byte.
Themis encodes textual passphrases in UTF-8 and uses the resulting byte string.

Since passphrases are typically shorter than keys and have less entropy (randomness),

> Secure Cell uses additional KDF rounds to preprocess passphrase data into an encryption key.

Reviewer: ```suggestion
Secure Cell uses additional PBKDF2 rounds to preprocess passphrase data into an encryption key.
```

let's say somewhere that we use PBKDF2?

Author: I've explicitly avoided mentioning PBKDF2 here since it's only an overview of keys and passphrases, not Secure Cell operation. I believe, intricacies of the KDF choice belong to Secure Cell description specifically. For one, it's not only PBKDF2 there, ZRTP-style KDF is also thrown into the mix. It will be easier to have KDF description in one place so that when a new KDF is added to Secure Cell then it's not necessary to locate and edit a bunch of other places as well. Passphrase content does not really depend on the specific KDF in use.

Reviewer: If not mention PBKD2 explicitly, maybe let's link to a page where we describe usage of PBKDF2?

Author: > maybe let's link to a page where we describe usage of PBKDF2?

As soon as it there. My idea was to cross-link Secure Cell spec with key/passphrase spec once it's actually ready.

Reviewer: We already have mentions of PBKDF2 in docs. I suggest to put a link on Secure Cell docs. https://docs.cossacklabs.com/themis/crypto-theory/cryptosystems/secure-cell/#key-derivation-functions

Author: > We already have mentions of PBKDF2 in docs. I suggest to put a link on Secure Cell docs.
> https://docs.cossacklabs.com/themis/crypto-theory/cryptosystems/secure-cell/#key-derivation-functions

Okay, I've added links to this section.

This slightly augments the security of the passphrase, making it harder to crack.
Currently, the KDF accounts for about 15 bits of additional entropy for passphrases.
This is not something groundbreaking, but at least it makes 4-digit PIN codes
and terrible passphrases a bit less terrible, just in case they are used.

### Generating passphrases

The recommended approach to generating passphrases is to generate a symmetric key of an adequate length,
then render it into a mnemonic which is easier for humans to remember.
For example, you can use a list of 2048 words to convert a 128-bit key into a sequence of 12 words.

{{< hint danger >}}
**Warning:**

> As for definitely **not recommended** approaches, avoid letting users choose their passphrases.
> Humans are rarely capable of generating cryptographically secure random sentences.

Reviewer: ```suggestion
As for definitely **not recommended** approaches, avoid letting users choose their cryptographic keys, as humans are rarely capable of generating cryptographically secure random sentences. 
```
I don't know where "avoid letting users choose their passphrases" is coming from. Memorizable secrets are foundation for many security controls, NIST guides have certain rules for handling memorizable secrets.

These sentences should be rewritten or removed.

Author: > I don't know where "avoid letting users choose their passphrases" is coming from.

From my head, obviously.

The stress here is on *"cryptographically secure"* (i.e., with comparable security to keys). Sure, humans are choosing their passphrases and ultimately it comes down to something kept in the head. However, without training and education, typical people are unlikely to choose a strong passphrase either, just like they are usually not able to generate keys of acceptable quality from their head.

Current Secure Cell KDF parameters are tweaked for *strong* passphrases. The slowdown of KDF contributes only about a dozen additional entropy bits and the passphrase is responsible for the rest. We recommend 256-bit keys, so my overall understanding is that the passphrases should have comparable strength. Or, well, at least not expect *majority* of entropy to come from the KDF alone and allow users to choose 4-digit PINs, claiming them to be as secure as the keys generated by Themis. For that, we'd need way more KDF rounds.

Anyways, if this is controversial and we don't have any advice to give here then I'm for removing this paragraph.

Reviewer: > then I'm for removing this paragraph.

or let's comment it out? maybe future-us will get back here and update this piece

Author: > or let's comment it out? maybe future-us will get back here and update this piece

I'm not a fan of leaving swathes of commented out code, or documentation. I thought about replacing it with a `<!-- TODO -->`, linking to this discussion. But after further consideration I don't really see anything of value to be put here. We already give an idea of how a passphrase can be generated and give links to NIST recommendations on the topic. I guess that should be enough. If future editors are really interested, they can `git blame` the file.

Reviewer: okay please do.

{{< /hint >}}

> There are no strict recommendations for a passphrase length since it mostly depends on human memory capabilities,
> but typically this is about 80 bits for offline passphrases and at least 128 bits for online use.
> Obviously, the longer the better.

Reviewer: ```suggestion
[NIST SP 800-63](https://www.nist.gov/itl/tig/projects/special-publication-800-63) suggests using 8 characters as minimum, and up to 64 characters for user-generated passphrases (see [NIST SP 800-63b, 5.1.1 Memorized Secrets](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-63b.pdf)).
```
```

Author: Well... It's contestable whether guidelines for *authentication* apply to secrets used to secure data. But okay, a reference is better than no reference.

Do not use general-purpose random number generators – typically found in “math” libraries –
unless they are explicitly designed for cryptographic purposes.
{{< /hint >}}

As of 2020,
[NIST recommends](https://www.keylength.com/en/4/)
symmetric keys to be at least 128 bits long for them to be future-proof.
Themis currently generates 256-bit keys by default (i.e., 32 bytes).

See also our [key management guide](/themis/crypto-theory/key-management/)
for advice on securing the key after you have generated it.

### Example

Here is a snippet illustrating how to generate a good symmetric key:

```go
package main

import (
	"crypto/rand" // Use a cryptograpic RNG.
	"fmt"
)

func main() {
	// Generate keys of sufficient length.
	const keyBytes = 32
	key := make([]byte, keyBytes)
	_, err := rand.Read(key)
	if err != nil {
		// Not handling RNG failure is a common critical error.
		// If errors are ignored, you might return a predictable key,
		// such as "all zeros" initial value of the byte array.
		panic("failed to generate a key")
	}
	fmt.Printf("✅ symmetric key: %x\n", key)
}
```

## Passphrases

Alternatively, Secure Cell can work with passphrases.

> From the cryptographic point of view, a passphrase is essentially a weird symmetric key.

Reviewer 2: I would write not "weird", but "weak". But its just IMO

Author: > I would write not "weird", but "weak" [symmetric key].

Mmm... It's not like passphrases are weak _per se_: given a sufficient length and randomness, they can be effectively as good as the keys. I guess I wanted to communicate here that passphrases have lower entropy per byte than keys, that's why they are "worse".

I'll think of a better wording. At the moment, I'm out of ideas. @vixentael?

Author: Well, if that's about entropy then I've just written *that* as is, without calling passphrases weird. Please give it a look.
